pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -0,0 +1,139 @@
+
+package source.v1;
+
+service UserDefinedSource {
+  // Read returns a stream of datum responses.
+  // The size of the returned ReadResponse is less than or equal to the num_records specified in ReadRequest.
+  rpc Read(ReadRequest) returns (stream ReadResponse);
+
+  // Ack acknowledges a list of datum offsets.
+  // It indicates that the datum stream has been processed by the source vertex.
+  rpc Ack(AckRequest) returns (AckResponse);
+
+  // Pending returns the number of pending records at the user defined source.
+  rpc Pending(google.protobuf.Empty) returns (PendingResponse);
+
+  // IsReady is the heartbeat endpoint for gRPC.
+  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
+}
+
+/**
+ * EventTime represents the event time of the datum.
+ */
+message EventTime {
+  google.protobuf.Timestamp event_time = 1;
+}
+
+/**
+ * ReadRequest is the request for reading datum stream from user defined source.
+ */
+message ReadRequest {
+  message Request {
+    // num_records is the number of records to read.
+    uint64 num_records = 1;
+  }
+  Request request = 1;
+}
+
+/**
+ * ReadResponse represents a list of datum response elements.
+ */
+message ReadResponse {
+  message Result {
+    // elements is a list of datum response elements.
+    repeated DatumResponse elements = 1;
+  }
+  Result result = 1;
+}
+
+/**
+ * DatumResponse represents a datum response element.
+ */
+message DatumResponse {
+  message Result {
+    // payload is the payload of the datum.
+    bytes payload = 1;
+    // offset is the offset information of the datum.
+    Offset offset = 2;
+    // keys is an optional list of keys associated with the datum.
+    // Key is the "key" attribute in (key,value) as in the map-reduce paradigm.
+    // We add this optional field to support the use case where the user defined source can provide keys for the datum.
+    // e.g. Kafka and Redis Stream message usually include information about the keys.
+    repeated string keys = 3;
+    // event_time is the time associated with each datum.
+    // We add this optional field to support the use case where the user defined source can provide event time for the datum.
+    // e.g. Kafka and Redis Stream message usually include information about the event time.
+    EventTime event_time = 4;
+  }
+  Result result = 1;
+}
+
+/**
+ * AckRequest is the request for acknowledging datum.
+ * It takes a list of offsets to be acknowledged.
+ */
+message AckRequest {
+  message Request {
+    repeated Offset offsets = 1;
+  }
+  Request request = 1;
+}
+
+/**
+ * AckResponse is the response for acknowledging datum.
+ */
+message AckResponse {
+  message Result {
+    // status is the status of the AckResponse.
+    AckStatus status = 1;
+    // non_acked_offsets is the list of offsets that are not acknowledged.
+    repeated Offset non_acked_offsets = 2;
+  }
+  Result result = 1;
+  /**
+   * AckStatus is the status of the AckResponse.
+   */
+  enum AckStatus {
```

Comment:
what will be the behavior for different AckStatus values? Should we make it either all successful or none? How will the source contract look like in the case of partically_acked? Could you please document those too?